processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -147,23 +135,14 @@ protected Integer addToFacts(
             makeColumnSelectorFactory(agg, rowSupplier, deserializeComplexMetrics)
         );
       }
-      final Integer rowIndex = indexIncrement.getAndIncrement();
-
-      concurrentSet(rowIndex, aggs);
-
-      // Last ditch sanity checks
-      if (numEntries.get() >= maxRowCount && !facts.containsKey(key)) {
+      if (numEntries.incrementAndGet() > maxRowCount) {
+        numEntries.decrementAndGet();
```

Comment:
incrementing followed by decrementing could lead to some nasty race conditions if anything relies on `size()` or `isEmpty()` in between those two calls.